x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -500,48 +355,167 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
 
     public static List<EsIndex> separateMappings(DataTypeRegistry typeRegistry, String javaRegex, String[] indexNames,
             Map<String, Map<String, FieldCapabilities>> fieldCaps, ImmutableOpenMap<String, List<AliasMetadata>> aliases) {
-        return buildIndices(typeRegistry, indexNames, javaRegex, fieldCaps, aliases, Function.identity(), (s, cap) -> null);
+        return buildIndices(typeRegistry, indexNames, javaRegex, fieldCaps, aliases, Function.identity(), false);
+    }
+
+    private static Map<String, QlFieldCapabilities> buildQlFieldCaps(String[] indexNames,
+            Map<String, Map<String, FieldCapabilities>> fieldCaps, ImmutableOpenMap<String, List<AliasMetadata>> indexToAliases,
+            boolean isMergedMapping) {
+        Map<String, QlFieldCapabilities> map = new HashMap<>();
+        Map<String, Set<String>> aliasToIndices = new HashMap<>(); // map with aliases and their list of indices
+        
+        if (indexToAliases != null) {
+            Iterator<ObjectObjectCursor<String, List<AliasMetadata>>> iter = indexToAliases.iterator();
+            while (iter.hasNext()) {
+                ObjectObjectCursor<String, List<AliasMetadata>> index = iter.next();
+                for (AliasMetadata aliasMetadata : index.value) {
+                    String aliasName = aliasMetadata.alias();
+                    aliasToIndices.putIfAbsent(aliasName, new HashSet<>());
+                    aliasToIndices.get(aliasName).add(index.key);
+                }
+            }
+        } else if (isMergedMapping) {
+            // single virtual alias named "*" to be used for a merged mappings
+            aliasToIndices.put("*", new HashSet<>(Arrays.asList(indexNames)));
+        }
```

Comment:
I find this bit confusing - what happens if the indexToAliases is null and isMergedMapping is false?
Since `isMergedMapping` is used only during the initialization of `indexToAliases`, I think it's easier to get rid of it and use it as part of `indexToAliases` creation which is created outside this method as well.

If I understand correctly, either there are aliases in which case the mapping back from alias to indices needs to be done or the mapping needs to be merged. Maybe cover the other branch by checking whether indexToAliases is null or making it empty or having a single entry, that of `*` essentially pointing all indices to the same virtual alias.

Reply:
Agree, it's confusing. It's an attempt at handling three different scenarios.
I've rewritten this part and removed `isMergedMapping` completely and `aliasToIndices` comes already built from the two places where this method is being called: merged mappings and separate mappings.